contrib/nitrpg/src/achievements.nit

```diff
@@ -0,0 +1,402 @@
+	end
+
+	redef fun to_json do
+		var json = super
+		json["id"] = id
+		json["name"] = name
+		json["desc"] = desc
+		json["reward"] = reward
+		return json
+	end
+end
+
+redef class Player
+
+	# Is `a` unlocked for this `Player`?
+	fun has_achievement(a: Achievement): Bool do
+		return load_achievement(a.id) != null
+	end
+
+	# Unlocks an achievement for this Player based on a GithubEvent.
+	#
+	# Register the achievement and adds the achievement reward to the player
+	# nitcoins.
+	#
+	# Do nothing is this player has already unlocked the achievement.
+	#
+	# TODO: add abstraction so achievements do not depend on GithubEvent.
+	fun unlock_achievement(a: Achievement, event: GithubEvent) do
+		if has_achievement(a) then return
+		nitcoins += a.reward
+		add_achievement(a)
+		trigger_unlock_event(a, event)
+	end
+
+	# Create a new event that marks the achievement unlocking.
+	fun trigger_unlock_event(achievement: Achievement, event: GithubEvent) do
+		var obj = new JsonObject
+		obj["player"] = name
+		obj["reward"] = achievement.reward
+		obj["achievement"] = achievement.id
+		obj["github_event"] = event.json
+		var ge = new GameEvent(game, "achievement_unlocked", obj)
+		add_event(ge)
+		game.add_event(ge)
+		achievement.add_event(ge)
+	end
+end
+
+# `GameReactor` dedicated to achievements unlocking.
+interface AchievementReactor
+	super GameReactor
+
+	# Unic ID of the achievement this reactor unlocks.
+	fun id: String is abstract
+
+	# Name of the achievement this reactor unlocks.
+	fun name: String is abstract
+
+	# Description of the achievement this reactor unlocks.
+	fun desc: String is abstract
+
+	# Amount of nitcoins rewarded for unlocking the achievement.
+	fun reward: Int is abstract
+
+	# Return a new instance of the achievement to unlock.
+	fun new_achievement(game: Game): Achievement do
+		var achievement = new Achievement(game, id, name, desc, reward)
+		game.add_achievement(achievement)
+		return achievement
+	end
+end
+
+#####################
+### Issues
+#####################
+
+# Unlock achievement after X issues.
+#
+# Used to factorize behavior.
+abstract class PlayerXIssues
+	super AchievementReactor
+
+	# Number of PR required to unlock the achievement.
+	var threshold: Int is noinit
+
+	redef fun react_event(game, event) do
+		if not event isa IssuesEvent then return
+		if not event.action == "opened" then return
+		var player = event.issue.user.player(game)
+		if player.stats["issues"] == threshold then
+			var a = new_achievement(game)
+			player.unlock_achievement(a, event)
+		end
+	end
+end
+
+# Player open his first issue.
+class Player1Issue
+	super PlayerXIssues
+
+	redef var id = "player_1_issue"
+	redef var name = "First complaint"
+	redef var desc = "Open your first issue."
+	redef var reward = 10
+	redef var threshold = 1
+end
+
+# Player open 100 issues.
+class Player100Issues
+	super PlayerXIssues
+
+	redef var id = "player_100_issues"
+	redef var name = "Hundred of complaints"
```

Comment:
Je suis pas sur du "of" ici, "Hundred complaints" serait plus grammaticalement correct, non ?
Sinon je peux proposer un nom un peu plus l'fun ? :) Style "Mature whiner" ou un truc de même ?